src/dcs/dcs.cc

```diff
@@ -1,168 +1,602 @@
   {
-    constexpr unsigned n = 8;
-    ASSERT((len - nread) >= 8);
-    auto rv = sscanf(buf + nread, "%08lx", &address);
-    ASSERT(rv == 1);
-    nread += n;
+	  constexpr unsigned n = 12;
+	  ASSERT((len - nread) >= n);
+	  expansion = std::string(buf + nread, n);
+	  nread += n;
   }
 
-  // YYDDDHHMMSS – Time the message arrived at the Wallops receive station.
-  // The day is represented as a three digit day of the year (julian day).
+  // 3.1.4.2 HDR_CRC32:= 4 bytes
   {
-    constexpr unsigned n = 11;
-    ASSERT((len - nread) >= n);
-    auto year = std::stoi(std::string(buf + nread, 2), nullptr, 10);
-    nread += 2;
-    auto day = std::stoi(std::string(buf + nread, 3), nullptr, 10);
-    nread += 3;
-    auto hour = std::stoi(std::string(buf + nread, 2), nullptr, 10);
-    nread += 2;
-    auto minute = std::stoi(std::string(buf + nread, 2), nullptr, 10);
-    nread += 2;
-    auto second = std::stoi(std::string(buf + nread, 2), nullptr, 10);
-    nread += 2;
-
-    struct tm tm;
-    memset(&tm, 0, sizeof(tm));
-
-    // The number of years since 1900.
-    tm.tm_year = 100 + year;
-
-    // Use offset at beginning of year and add parts
-    time.tv_sec = mktime(&tm) + ((((day * 24) + hour) * 60) + minute) * 60 + second;
-    time.tv_nsec = 0;
-  }
-
-  // 1 character failure code
-  {
-    constexpr unsigned n = 1;
-    ASSERT((len - nread) >= n);
-    failure = buf[nread];
-    nread += n;
+	  constexpr unsigned n = 4;
+	  ASSERT((len - nread) >= n);
+	  memcpy(&headerCRC, &buf[nread], n);
+	  nread += n;
   }
 
-  // 2 decimal digit signal strength
+  // 3.5 FILE_CRC32:= 4 bytes
   {
-    constexpr unsigned n = 2;
-    ASSERT((len - nread) >= n);
-    signalStrength = std::stoi(std::string(buf + nread, 2));
-    nread += n;
+	constexpr unsigned n = 4;
+	ASSERT((len - nread) >= n);
+	memcpy(&fileCRC, &buf[length - n], n); // CRC is at the end of the DCS payload
+	nread += n;
   }
 
-  // 2 decimal digit frequency offset
-  {
-    constexpr unsigned n = 2;
-    ASSERT((len - nread) >= n);
-    // This can be +A or -A; don't know what it means...
-    if ((buf[nread] == '+' || buf[nread] == '-') && buf[nread+1] == 'A') {
-      frequencyOffset = 0;
-    } else {
-      frequencyOffset = std::stoi(std::string(buf + nread, n));
+	
+  return nread;
+}
+
+int DCPData::readFrom(const char* buf, size_t len) {
+  size_t nread = 64; // Start position of message block after file header
+  uint16_t pos = 0;
+
+
+  while (nread < len - 4) { // 4 for ending header CRC32 per spec
+    blocks.resize(pos + 1);
+
+    // 1 byte contains blocks ID
+    {
+      constexpr unsigned n = 1;
+      ASSERT((len - nread) >= n);
+      memcpy(&blocks[pos].blockID, &buf[nread], n);
+      nread += n;
+    }
+
+    // Skip over Missed Messages Blocks (blockID == 2 as of January 21, 2021) 
+    if (blocks[pos].blockID != 1) {
+      nread += 28; // Size of Missed Message Block (2(blocks len) + 24(header) + 2(CRC16))
```

Comment:
Skipping over 28 bytes is correct for messages with blockID=2, but it would be better to read the next 2 bytes to calculate the block length in order to figure out exactly how many bytes to skip